File: 킬링캠프/1주 차 - 완전탐색, 백트래킹, 이분탐색/practice/Word Search/solution.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        # 상수
        DIR = [[-1, 0], [0, 1], [0, -1], [1, 0]]  # 상, 우, 좌, 하
        M = len(board)
        N = len(board[0])

        # 메서드
        def search(row, col, visited, step):
            if step == len(word):
                return True
            
            for dx, dy in DIR:
                nextRow = row + dx
                nextCol = col + dy

                if not isSafe(nextRow, nextCol) or visited[nextRow][nextCol] or board[nextRow][nextCol] != word[step]:
                    continue

                visited[nextRow][nextCol] = True
                if search(nextRow, nextCol, visited, step + 1):
                    return True
                
                visited[nextRow][nextCol] = False
            
            return False
        
        def isSafe(row, col):
            return row >= 0 and row < M and col >= 0 and col < N

        # 메인 로직
        for row in range(M):
            for col in range(N):
                if board[row][col] != word[0]:
                    continue
                
                visited = [[False for _ in range(N)] for _ in range(M)]
                visited[row][col] = True
                if search(row, col, visited, 1):
                    return True
        
        return False

File: 킬링캠프/1주 차 - 완전탐색, 백트래킹, 이분탐색/practice/Word Search/Solution.java
import java.util.*;

class Solution {

    // 북, 동, 서, 남
    private static final int[][] STEP = new int[][]{{-1, 0}, {0, 1}, {0, -1}, {1, 0}};
    private static final char VISITED = '.';

    public boolean exist(char[][] board, String word) {
        for (int row = 0; row < board.length; row++) {
            for (int col = 0; col < board[row].length; col++) {
                if (board[row][col] != word.charAt(0)) {
                    continue;
                }

                board[row][col] = VISITED;
                if (backTracking(board, word, row, col, 1)) {
                    return true;
                }

                board[row][col] = word.charAt(0);
            }
        }

        return false;
    }

    private boolean backTracking(char[][] board, String word, int row, int col, int index) {
        if (index == word.length()) {
            return true;
        }

        for (int i = 0; i < 4; i++) {
            int nextRow = row + STEP[i][0];
            int nextCol = col + STEP[i][1];

            if (!isSafe(board, nextRow, nextCol) || board[nextRow][nextCol] != word.charAt(index)) {
                continue;
            }

            board[nextRow][nextCol] = VISITED;
            if (backTracking(board, word, nextRow, nextCol, index + 1)) {
                return true;
            }

            board[nextRow][nextCol] = word.charAt(index);
        }

        return false;
    }

    private boolean isSafe(char[][] board, int row, int col) {
        if (row >= 0 && row < board.length && col >= 0 && col < board[row].length && board[row][col] != VISITED) {
            return true;
        }

        return false;
    }
}
